File: modules/vescape-core/ios/alerts/CustomAppSounds.swift
import AVFoundation
import Foundation

/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/alerts/CustomAppSounds.kt
internal enum CustomAppSounds {
  private static let cues: Set<String> = ["on", "off", "error", "created", "join"]
  private static let maxBytes = 2_000_000
  private static let maxDuration = 15.0
  private static let lock = NSRecursiveLock()

  private struct Pack: Codable {
    var id: String
    var name: String
    var sounds: [String: String]
  }

  private static var directory: URL {
    let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
    return base.appendingPathComponent("custom-app-sounds", isDirectory: true)
  }
  private static var manifest: URL { directory.appendingPathComponent("packs.json") }
  private static func read() -> [Pack] {
    // intentional-suppression: absent or unreadable pack manifest means no custom packs; Classic remains usable
    guard let data = try? Data(contentsOf: manifest) else { return [] }
    // intentional-suppression: corrupt pack metadata falls back to an empty list and Classic playback
    return (try? JSONDecoder().decode([Pack].self, from: data)) ?? []
  }
  private static func save(_ packs: [Pack]) throws {
    try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    try JSONEncoder().encode(packs).write(to: manifest, options: .atomic)
  }
  private static func validatedName(_ name: String) throws -> String {
    let clean = name.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !clean.isEmpty, clean.count <= 60 else { throw error("Name must be 1–60 characters") }
    return clean
  }
  private static func error(_ message: String) -> NSError {
    NSError(domain: "CustomAppSounds", code: 1, userInfo: [NSLocalizedDescriptionKey: message])
  }
  static func list() -> [[String: Any]] {
    lock.lock(); defer { lock.unlock() }
    return read().map { ["id": $0.id, "name": $0.name, "sounds": $0.sounds] }
  }
  static func exists(_ id: String) -> Bool {
    lock.lock(); defer { lock.unlock() }
    return read().contains { $0.id == id }
  }
  static func create(_ name: String) throws -> [[String: Any]] {
    lock.lock(); defer { lock.unlock() }
    var packs = read()
    packs.append(Pack(id: UUID().uuidString.lowercased(), name: try validatedName(name), sounds: [:]))
    try save(packs)
    return list()
  }
  static func rename(_ id: String, name: String) throws {
    lock.lock(); defer { lock.unlock() }
    var packs = read()
    guard let index = packs.firstIndex(where: { $0.id == id }) else { throw error("Sound pack missing") }
    packs[index].name = try validatedName(name)
    try save(packs)
  }
  static func remove(_ id: String, cue: String) throws {
    lock.lock(); defer { lock.unlock() }
    guard cues.contains(cue) else { throw error("Unknown sound cue") }
    var packs = read()
    guard let index = packs.firstIndex(where: { $0.id == id }) else { throw error("Sound pack missing") }
    let old = packs[index].sounds.removeValue(forKey: cue)
    try save(packs)
    // intentional-suppression: orphaned old audio may be cleaned later; saved assignment is already current
    if let old { try? FileManager.default.removeItem(at: directory.appendingPathComponent(old)) }
  }
  static func delete(_ id: String) throws {
    lock.lock(); defer { lock.unlock() }
    var packs = read()
    guard let index = packs.firstIndex(where: { $0.id == id }) else { throw error("Sound pack missing") }
    let old = packs.remove(at: index)
    try save(packs)
    // intentional-suppression: deleting metadata is authoritative; leftover files are inert
    for name in old.sounds.values { try? FileManager.default.removeItem(at: directory.appendingPathComponent(name)) }
  }
  static func `import`(_ id: String, cue: String, uri: String) throws {
    lock.lock(); defer { lock.unlock() }
    guard cues.contains(cue) else { throw error("Unknown sound cue") }
    var packs = read()
    guard let index = packs.firstIndex(where: { $0.id == id }) else { throw error("Sound pack missing") }
    guard let source = URL(string: uri), source.isFileURL else { throw error("Could not open audio file") }
    try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    let name = "\(UUID().uuidString.lowercased()).wav"
    let destination = directory.appendingPathComponent(name)
    do {
      try FileManager.default.copyItem(at: source, to: destination)
      let attrs = try FileManager.default.attributesOfItem(atPath: destination.path)
      guard let size = attrs[.size] as? Int, size > 0, size <= maxBytes else { throw error("WAV must be under 2 MB") }
      let handle = try FileHandle(forReadingFrom: destination)
      let header = handle.readData(ofLength: 12)
      try handle.close()
      guard header.count == 12, String(data: header.prefix(4), encoding: .ascii) == "RIFF",
        String(data: header.suffix(4), encoding: .ascii) == "WAVE" else { throw error("Only WAV files are supported") }
      let audio: AVAudioFile
      do { audio = try AVAudioFile(forReading: destination) }
      catch { throw Self.error("WAV audio could not be decoded") }
      guard audio.processingFormat.sampleRate >= 8_000,
        audio.processingFormat.sampleRate <= 192_000,
        audio.length <= 2_880_000 else { throw error("Unsupported WAV audio format") }
      let duration = Double(audio.length) / audio.processingFormat.sampleRate
      guard duration > 0, duration <= maxDuration else { throw error("Audio must be 15 seconds or shorter") }
      guard let buffer = AVAudioPCMBuffer(
        pcmFormat: audio.processingFormat,
        frameCapacity: AVAudioFrameCount(audio.length)
      ) else { throw error("WAV audio could not be decoded") }
      do { try audio.read(into: buffer) }
      catch { throw Self.error("WAV audio could not be decoded") }
      guard buffer.frameLength > 0 else { throw error("WAV audio could not be decoded") }
      let old = packs[index].sounds.updateValue(name, forKey: cue)
      try save(packs)
      // intentional-suppression: orphaned old audio may be cleaned later; saved assignment is already current
    if let old { try? FileManager.default.removeItem(at: directory.appendingPathComponent(old)) }
    } catch {
      // intentional-suppression: cleanup after failed import is best effort; original error is rethrown
      try? FileManager.default.removeItem(at: destination)
      throw error
    }
  }
  static func file(_ id: String, cue: String) -> URL? {
    lock.lock(); defer { lock.unlock() }
    guard cues.contains(cue), let name = read().first(where: { $0.id == id })?.sounds[cue],
      name.range(of: "^[0-9a-f-]{36}\\.wav$", options: .regularExpression) != nil else { return nil }
    let url = directory.appendingPathComponent(name)
    // intentional-suppression: missing file or unreadable attributes use Classic playback
    guard let size = try? FileManager.default.attributesOfItem(atPath: url.path)[.size] as? Int,
      size > 0, size <= maxBytes else { return nil }
    return url
  }

  static func archiveEntries() -> [String: Data] {
    lock.lock(); defer { lock.unlock() }
    var entries: [String: Data] = [:]
    // intentional-suppression: no readable manifest means backup contains no custom packs
    if let data = try? Data(contentsOf: manifest) { entries["packs.json"] = data }
    for pack in read() {
      for name in pack.sounds.values {
        // intentional-suppression: unreadable audio is omitted and its assignment is pruned on restore
        if let data = try? Data(contentsOf: directory.appendingPathComponent(name)), data.count <= maxBytes {
          entries[name] = data
        }
      }
    }
    return entries
  }

  static func validateBackup(_ entries: [String: Data]) throws {
    guard let data = entries["packs.json"] else { return }
    let packs = try JSONDecoder().decode([Pack].self, from: data)
    guard packs.allSatisfy({ !$0.id.isEmpty && !$0.name.isEmpty }) else {
      throw error("Invalid sound pack")
    }
  }

  static func replaceFromBackup(_ entries: [String: Data]) throws {
    lock.lock(); defer { lock.unlock() }
    let decoder = JSONDecoder()
    var packs = try entries["packs.json"].map { try decoder.decode([Pack].self, from: $0) } ?? []
    for index in packs.indices {
      packs[index].sounds = packs[index].sounds.filter { cue, name in
        cues.contains(cue) && name.range(of: "^[0-9a-f-]{36}\\.wav$", options: .regularExpression) != nil &&
          (entries[name]?.count ?? 0) > 0 && (entries[name]?.count ?? 0) <= maxBytes
      }
    }
    let old = directory.deletingLastPathComponent().appendingPathComponent("custom-app-sounds-old")
    // intentional-suppression: stale restore rollback directory is disposable
    try? FileManager.default.removeItem(at: old)
    if FileManager.default.fileExists(atPath: directory.path) { try FileManager.default.moveItem(at: directory, to: old) }
    do {
      try save(packs)
      for pack in packs {
        for name in pack.sounds.values {
          if let data = entries[name] { try data.write(to: directory.appendingPathComponent(name), options: .atomic) }
        }
      }
      // intentional-suppression: stale restore rollback directory is disposable
    try? FileManager.default.removeItem(at: old)
    } catch {
      // intentional-suppression: rollback cleanup is best effort; original restore error is rethrown
      try? FileManager.default.removeItem(at: directory)
      // intentional-suppression: rollback attempt is best effort; original restore error is rethrown
      if FileManager.default.fileExists(atPath: old.path) { try? FileManager.default.moveItem(at: old, to: directory) }
      throw error
    }
  }
}

File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/alerts/CustomAppSounds.kt
package expo.modules.vescapecore.alerts

import android.content.Context
import android.media.AudioAttributes
import android.media.MediaMetadataRetriever
import android.media.MediaPlayer
import android.net.Uri
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.util.UUID

/** @parity /modules/vescape-core/ios/alerts/CustomAppSounds.swift */
internal object CustomAppSounds {
  private val cues = setOf("on", "off", "error", "created", "join")
  private const val maxBytes = 2_000_000L
  private const val maxDurationMs = 15_000L
  private fun directory(context: Context) = File(context.filesDir, "custom-app-sounds").apply { mkdirs() }
  private fun manifest(context: Context) = File(directory(context), "packs.json")
  private fun read(context: Context): JSONArray = try {
    JSONArray(manifest(context).readText())
  } catch (_: Exception) { JSONArray() }
  private fun save(context: Context, packs: JSONArray) {
    val target = manifest(context)
    val temp = File(target.parentFile, "packs.json.tmp")
    temp.writeText(packs.toString())
    check(temp.renameTo(target)) { "Could not save sound packs" }
  }
  private fun find(packs: JSONArray, id: String): Pair<Int, JSONObject>? {
    for (i in 0 until packs.length()) {
      val pack = packs.getJSONObject(i)
      if (pack.optString("id") == id) return i to pack
    }
    return null
  }
  @Synchronized fun list(context: Context): List<Map<String, Any>> {
    val packs = read(context)
    return (0 until packs.length()).map { i ->
      val pack = packs.getJSONObject(i)
      val sounds = pack.optJSONObject("sounds") ?: JSONObject()
      mapOf("id" to pack.getString("id"), "name" to pack.getString("name"),
        "sounds" to cues.filter { sounds.has(it) }.associateWith { sounds.getString(it) })
    }
  }
  @Synchronized fun exists(context: Context, id: String): Boolean = find(read(context), id) != null
  @Synchronized fun create(context: Context, name: String): List<Map<String, Any>> {
    val clean = name.trim()
    require(clean.isNotEmpty() && clean.length <= 60) { "Name must be 1–60 characters" }
    val packs = read(context)
    packs.put(JSONObject().put("id", UUID.randomUUID().toString()).put("name", clean).put("sounds", JSONObject()))
    save(context, packs)
    return list(context)
  }
  @Synchronized fun rename(context: Context, id: String, name: String) {
    val clean = name.trim()
    require(clean.isNotEmpty() && clean.length <= 60) { "Name must be 1–60 characters" }
    val packs = read(context)
    val pack = requireNotNull(find(packs, id)) { "Sound pack missing" }.second
    pack.put("name", clean)
    save(context, packs)
  }
  @Synchronized fun remove(context: Context, id: String, cue: String) {
    require(cue in cues) { "Unknown sound cue" }
    val packs = read(context)
    val pack = requireNotNull(find(packs, id)) { "Sound pack missing" }.second
    val old = pack.getJSONObject("sounds").optString(cue)
    pack.getJSONObject("sounds").remove(cue)
    save(context, packs)
    if (old.isNotEmpty()) File(directory(context), old).delete()
  }
  @Synchronized fun delete(context: Context, id: String) {
    val packs = read(context)
    val (index, pack) = requireNotNull(find(packs, id)) { "Sound pack missing" }
    val sounds = pack.getJSONObject("sounds")
    val names = cues.map { sounds.optString(it) }.filter { it.isNotEmpty() }
    packs.remove(index)
    save(context, packs)
    names.forEach { File(directory(context), it).delete() }
  }
  @Synchronized fun import(context: Context, id: String, cue: String, uri: String) {
    require(cue in cues) { "Unknown sound cue" }
    val packs = read(context)
    val pack = requireNotNull(find(packs, id)) { "Sound pack missing" }.second
    val name = "${UUID.randomUUID()}.wav"
    val target = File(directory(context), name)
    try {
      context.contentResolver.openInputStream(Uri.parse(uri)).use { input ->
        requireNotNull(input) { "Could not open audio file" }
        target.outputStream().use { output ->
          val buffer = ByteArray(8192)
          var bytes = 0L
          while (true) {
            val count = input.read(buffer)
            if (count < 0) break
            bytes += count
            require(bytes <= maxBytes) { "WAV must be under 2 MB" }
            output.write(buffer, 0, count)
          }
          require(bytes > 0) { "Empty WAV file" }
        }
      }
      val header = ByteArray(12)
      target.inputStream().use { require(it.read(header) == 12) { "Invalid WAV file" } }
      require(String(header, 0, 4) == "RIFF" && String(header, 8, 4) == "WAVE") { "Only WAV files are supported" }
      val retriever = MediaMetadataRetriever()
      try {
        try { retriever.setDataSource(target.absolutePath) }
        catch (_: Exception) { throw IllegalArgumentException("WAV audio could not be decoded") }
        val duration = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)?.toLongOrNull()
        require(duration != null && duration in 1..maxDurationMs) { "Audio must be 15 seconds or shorter" }
      } finally { retriever.release() }
      val decoder = MediaPlayer()
      try {
        decoder.setDataSource(target.absolutePath)
        decoder.prepare()
      } catch (_: Exception) {
        throw IllegalArgumentException("WAV audio could not be decoded")
      } finally { decoder.release() }
      val old = pack.getJSONObject("sounds").optString(cue)
      pack.getJSONObject("sounds").put(cue, name)
      save(context, packs)
      if (old.isNotEmpty()) File(directory(context), old).delete()
    } catch (error: Exception) {
      target.delete()
      throw error
    }
  }
  @Synchronized fun file(context: Context, id: String, cue: String): File? {
    if (cue !in cues) return null
    val pack = find(read(context), id)?.second ?: return null
    val name = pack.optJSONObject("sounds")?.optString(cue).orEmpty()
    if (!Regex("[0-9a-fA-F-]{36}\\.wav").matches(name)) return null
    return File(directory(context), name).takeIf { it.isFile && it.length() in 1..maxBytes }
  }
  fun play(context: Context, id: String, cue: String, source: String, onFailure: () -> Unit): Boolean {
    val sound = file(context, id, cue) ?: return false
    val player = MediaPlayer()
    return try {
      player.setAudioAttributes(audioAttributes(source, AudioAttributes.CONTENT_TYPE_SONIFICATION))
      player.setDataSource(sound.absolutePath)
      player.setOnPreparedListener {
        try { it.start() }
        catch (_: Exception) { it.release(); onFailure() }
      }
      player.setOnCompletionListener { it.release() }
      player.setOnErrorListener { p, _, _ -> p.release(); onFailure(); true }
      player.prepareAsync()
      true
    } catch (_: Exception) { player.release(); false }
  }
  /** Validate and prune a candidate before its database is installed. */
  @Synchronized fun validateBackup(staged: File) {
    val manifestFile = File(staged, "packs.json")
    if (manifestFile.exists()) {
      val packs = JSONArray(manifestFile.readText())
      for (i in 0 until packs.length()) {
        val pack = packs.getJSONObject(i)
        require(pack.getString("id").isNotEmpty() && pack.getString("name").isNotEmpty()) { "Invalid sound pack" }
        val sounds = pack.getJSONObject("sounds")
        for (cue in cues) {
          val name = sounds.optString(cue)
          if (name.isNotEmpty() && (!Regex("[0-9a-fA-F-]{36}\\.wav").matches(name) ||
                !File(staged, name).isFile)) sounds.remove(cue)
        }
      }
      manifestFile.writeText(packs.toString())
    }
  }

  /** Old backups contain no sound entries. A restore then clears packs and resets selection on read. */
  @Synchronized fun replaceFromBackup(context: Context, staged: File) {
    val destination = File(context.filesDir, "custom-app-sounds")
    val previous = File(context.filesDir, "custom-app-sounds-old")
    previous.deleteRecursively()
    if (destination.exists()) check(destination.renameTo(previous)) { "Could not stage old sound packs" }
    try {
      if (staged.exists()) check(staged.copyRecursively(destination, overwrite = true)) { "Could not restore sound files" }
      previous.deleteRecursively()
    } catch (error: Exception) {
      destination.deleteRecursively()
      if (previous.exists()) previous.renameTo(destination)
      throw error
    }
  }
}
